File: scripts/test-runner.ts
// test-runner.ts
import { $ } from "bun"
import { parseArgs } from "util"
import { resolve, join, dirname } from "path"
import { existsSync, readFileSync, writeFileSync, mkdtempSync } from "fs"
import { tmpdir } from "os"
import { globSync } from "glob"

type OpenMode = "claude" | "file"

interface RunTestsParams {
  filePath: string
  command?: string
  cwd?: string
  open?: OpenMode
}

function findClosestPackageJson(from: string): string | null {
  let current = existsSync(from) && !Bun.file(from).name?.includes(".")
    ? from
    : dirname(from)

  while (current !== dirname(current)) {
    const candidate = join(current, "package.json")
    if (existsSync(candidate)) return candidate
    current = dirname(current)
  }
  return null
}

function findTypesFiles(dir: string): string[] {
  return [
    ...globSync(join(dir, "types.ts")),
    ...globSync(join(dir, "*.types.ts")),
  ]
}

function extractErroringFiles(output: string, packageDir: string): string[] {
  const patterns = [
    /(?:FAIL|ERROR|error)\s+(.+?\.[tj]sx?)/g,
    /([^\s]+?\.[tj]sx?)[\s:(]\d+/g,
    /at\s+.*?\((.+?\.[tj]sx?):\d+:\d+\)/g,
  ]

  const files = new Set<string>()

  for (const pattern of patterns) {
    for (const match of output.matchAll(pattern)) {
      const raw = match[1].trim()
      const fromPkg = resolve(packageDir, raw)
      const absolute = resolve(raw)

      if (existsSync(fromPkg)) files.add(fromPkg)
      else if (existsSync(absolute)) files.add(absolute)
    }
  }

  return [...files].sort()
}

function readFileSafe(path: string): string {
  try {
    return readFileSync(path, "utf-8")
  } catch (e) {
    return `[Could not read: ${e}]`
  }
}

async function openAsFile(prompt: string) {
  const tmpDir = mkdtempSync(join(tmpdir(), "test-runner-"))
  const tmpFile = join(tmpDir, "prompt.md")
  writeFileSync(tmpFile, prompt, "utf-8")
  console.log(`\n📄 Prompt written to: ${tmpFile}`)
  await $`open ${tmpFile}`.quiet()
}

async function runTests({
  filePath,
  command = "bun test",
  open = "claude",
}: RunTestsParams) {
  const resolved = resolve(filePath)
  const pkgJson = findClosestPackageJson(resolved)
  const discoveredDir = pkgJson ? dirname(pkgJson) : null
  const workDir = discoveredDir

  if (!workDir) {
    console.error(`No package.json found for ${filePath} and no --cwd provided`)
    process.exit(1)
  }

  console.log(`📁 Working directory: ${workDir}`)
  console.log(`🧪 Running: ${command}\n`)

  const [cmd, ...args] = command.split(" ")
  const proc = Bun.spawnSync(args.length ? [cmd, ...args] : [cmd], {
    cwd: workDir,
    stdout: "pipe",
    stderr: "pipe",
    env: { ...process.env, FORCE_COLOR: "0" },
  })

  const stdout = proc.stdout.toString()
  const stderr = proc.stderr.toString()
  const combinedOutput = `${stdout}\n${stderr}`.trim()

  if (proc.exitCode === 0) {
    console.log("✅ All tests passed!\n")
    console.log(combinedOutput)
    return
  }

  // console.log("❌ Tests failed.\n")
  // console.log(combinedOutput)

  const erroringFiles = extractErroringFiles(combinedOutput, workDir)
  const typesFiles = findTypesFiles(workDir)
  const allFiles = [...new Set([...erroringFiles, ...typesFiles])]

  // console.log(`\nCollected ${erroringFiles.length} erroring file(s) and ${typesFiles.length} types file(s)`)
  // for (const f of allFiles) console.log(`  → ${f}`)

  const fileContents = allFiles
    .map((f) => `--- ${f} ---\n${readFileSafe(f)}`)
    .join("\n\n")

  const prompt = `The following tests failed. Please help me fix them.

## Test Output
\`\`\`
${combinedOutput}
\`\`\`

## Relevant Files
`

  writeFileSync('/tmp/abc.txt', prompt)
}

// --- CLI ---


await runTests({
  filePath: "/home/kdog3682/projects/paladin/packages/conversation/src/tests/diverse-imports-conversation.test.ts",
  command: 'bun test:analyze',
  open: 'file'
})

File: scripts/test-runner.py
# test-runner.py
import sys
import os
import re
import subprocess
import webbrowser
import urllib.parse
from pathlib import Path


def find_closest_package_json(file_path: str) -> Path | None:
    current = Path(file_path).resolve()
    if current.is_file():
        current = current.parent

    while current != current.parent:
        candidate = current / "package.json"
        if candidate.exists():
            return candidate
        current = current.parent

    return None


def find_types_files(package_dir: Path) -> list[Path]:
    return sorted(package_dir.glob("types.ts")) + sorted(package_dir.glob("*.types.ts"))


def extract_erroring_files(output: str, package_dir: Path) -> list[Path]:
    patterns = [
        r'(?:FAIL|ERROR|error)\s+(.+?\.[tj]sx?)',       # FAIL src/foo.ts
        r'([^\s]+?\.[tj]sx?)[\s:(]\d+',                   # src/foo.ts:10 or src/foo.ts(10,5)
        r'at\s+.*?\((.+?\.[tj]sx?):\d+:\d+\)',            # at Function (src/foo.ts:10:5)
    ]

    files = set()
    for pattern in patterns:
        for match in re.finditer(pattern, output):
            file_path = match.group(1).strip()
            resolved = (package_dir / file_path).resolve()
            if resolved.exists():
                files.add(resolved)
            elif Path(file_path).resolve().exists():
                files.add(Path(file_path).resolve())

    return sorted(files)


def read_file_contents(files: list[Path]) -> str:
    sections = []
    for f in files:
        try:
            content = f.read_text()
            sections.append(f"--- {f} ---\n{content}")
        except Exception as e:
            sections.append(f"--- {f} ---\n[Could not read: {e}]")
    return "\n\n".join(sections)


def main(file_path):
    pkg_json = find_closest_package_json(file_path)

    if not pkg_json:
        print(f"No package.json found for {file_path}")
        sys.exit(1)

    package_dir = pkg_json.parent
    print(f"Found package.json at: {pkg_json}")
    print(f"Running tests in: {package_dir}\n")

    result = subprocess.run(
        ["bun", "test"],
        cwd=package_dir,
        capture_output=True,
        text=True,
    )

    combined_output = result.stdout + "\n" + result.stderr

    if result.returncode == 0:
        print("✅ All tests passed!")
        print(combined_output)
        sys.exit(0)

    print("❌ Tests failed.\n")
    print(combined_output)

    # Gather offending files
    erroring_files = extract_erroring_files(combined_output, package_dir)
    types_files = find_types_files(package_dir)

    all_files = list(dict.fromkeys(erroring_files + types_files))  # dedupe, preserve order

    print(f"\nCollected {len(erroring_files)} erroring file(s) and {len(types_files)} types file(s)")
    for f in all_files:
        print(f"  → {f}")

    # Build the prompt
    file_contents = read_file_contents(all_files)
    prompt = f"""The following bun tests failed. Please help me fix them.

## Test Output
```
{combined_output.strip()}
```

## Relevant Files
{file_contents}
"""

    encoded = urllib.parse.quote(prompt, safe="")
    url = f"https://claude.ai/new?q={encoded}"

    # URL length sanity check — browsers cap around 2MB for most
    if len(url) > 1_500_000:
        print("\n⚠️  Payload too large for URL. Writing to temp file instead.")
        tmp = Path("/tmp/test-runner-prompt.md")
        tmp.write_text(prompt)
        print(f"Prompt saved to: {tmp}")
    else:
        print("\nOpening Claude with context...")
        webbrowser.open(url)


if __name__ == "__main__":
    main("/home/kdog3682/projects/paladin/packages/conversation/src/analyze.ts")
